File: framework/languages/dotnet/e2e/AutomaticTurnDispatch/Server/Session/SessionHostFactory.cs
using Microsoft.Extensions.Configuration;

using Systems.Zlink;
using AutomaticTurnDispatch.Server.Session.Support;
using AutomaticTurnDispatch.Shared;
using Zlink.Framework.AspNetCore;
using Zlink.Framework.Locations.Redis;
using Zlink.Framework.Contracts.Channels;
using Zlink.Framework.Contracts.Configuration;
using Zlink.Framework.Contracts.Dispatch;
using Zlink.Framework.Contracts.Errors;
using Zlink.Framework.Contracts.Actors;
using Zlink.Framework.E2E.Diagnostics;
using SessionServerOptions = AutomaticTurnDispatch.Server.Session.Support.SessionOptions;
using AwaitStreamSession = AutomaticTurnDispatch.Server.Session.Support.AwaitSession;

namespace AutomaticTurnDispatch.Server.Session;

internal static class SessionHostFactory
{
    public static WebApplication Create(string[] args)
    {
        var options = SessionServerOptions.Parse(args);
        Directory.CreateDirectory(options.LogDir);

        var builder = WebApplication.CreateBuilder(args);
        builder.Configuration.Sources.Clear();
        builder.Configuration.AddInMemoryCollection();
        builder.Logging.ClearProviders();
        builder.Logging.AddSimpleConsole(console =>
        {
            console.SingleLine = true;
            console.TimestampFormat = "HH:mm:ss.fff ";
        });
        builder.WebHost.UseUrls(options.HttpUrl);
        builder.Services.AddSingleton(new EvidenceStore(options.Rid, options.EvidenceFile));
        builder.Services.AddSingleton(new NodeOptions(options.Rid));
        builder.Services.AddSingleton(new E2eMessageFlowListener(
            Path.Combine(options.LogDir, $"{options.Rid}-flow.log"),
            options.Rid));
        builder.Services.AddZLinkFramework(framework =>
        {
            //  This E2E host is not started inside a memory-limited
            //  container. Supply a deterministic finite limit so the
            //  default Auto HWM contract does not depend on the host.
            framework.ConfigureInboundDispatch().ProcessMemoryLimitBytes =
                1UL * 1024 * 1024 * 1024;
            framework.AddLocationStore(new ZLinkRedisLocationStore(redis => { redis.ConnectionString = options.RedisEndpoint; redis.KeyPrefix = options.RedisKeyPrefix; }));
            framework.AddRelocationStore(new ZLinkRedisRelocationStore(redis => { redis.ConnectionString = options.RedisEndpoint; redis.KeyPrefix = $"{options.RedisKeyPrefix}:relocation"; }));
            framework.AddHandlersFromAssemblyOf(typeof(Program));
            framework.ConfigureDispatch().Diagnostics
                .SetLevel(ZLinkDiagnosticsLevel.Normal);
            var controlMesh = framework.AddRouteMesh(AutomaticTurnDispatchNames.ControlChannel)
                .Listen(options.ControlEndpoint)
                .SetRoutingIdPrefix(options.Rid);
            controlMesh.Channel(AutomaticTurnDispatchNames.ControlChannel).Client();
            var spotRouteMesh = framework.AddRouteMesh(AutomaticTurnDispatchNames.SpotRouteChannel)
                .Listen("tcp://127.0.0.1:0")
                .SetRoutingIdPrefix(options.Rid);
            spotRouteMesh.Channel(AutomaticTurnDispatchNames.SpotRouteChannel).Server().SetWeight(0);
            var mesh23 = framework.AddRouteMesh(AutomaticTurnDispatchNames.SpotChannel)
                .Listen(options.SpotRouterEndpoint)
                .SetRoutingIdPrefix(options.Rid);
            // This node serves the RouteMesh channel, so it is not outbound-only.
            // Object Server includes the Object Client operations used by the session.
            mesh23.Objects().Server();
            mesh23.Channel(AutomaticTurnDispatchNames.SpotChannel).Server();
            framework.AddStreamNode(AutomaticTurnDispatchNames.StreamNode)
                .Bind(options.StreamEndpoint)
                .EnableActorDispatch()
                .AddSession<AwaitStreamSession>();
        });

        var app = builder.Build();
        app.MapGet("/health", () => Results.Ok(new { status = "ready", role = "session", options.Rid }));
        app.MapGet("/topology/ready", async (
            string meshName,
            string rid,
            IZLinkRouteMeshRuntime runtime,
            IZLinkRouteClient routes,
            CancellationToken cancellationToken) =>
        {
            var snapshot = runtime.GetStatus(meshName);
            var snapshotReady = snapshot.Peers.Any(peer =>
                peer.State == ZLinkPeerState.Ready
                && (string.Equals(
                        peer.NodeRid.ToString(),
                        rid,
                        StringComparison.Ordinal)
                    || peer.NodeRid.ToString().StartsWith(
                        $"{rid}-",
                        StringComparison.Ordinal)));
            if (!snapshotReady
                || !string.Equals(
                    meshName,
                    AutomaticTurnDispatchNames.ControlChannel,
                    StringComparison.Ordinal))
                return Results.Ok(new { ready = snapshotReady });

            try
            {
                var target = snapshot.Peers
                    .Where(peer => peer.State == ZLinkPeerState.Ready)
                    .Select(peer => peer.NodeRid)
                    .First(peer => peer.ToString().StartsWith(
                        $"{rid}-",
                        StringComparison.Ordinal));
                _ = await routes.RequestToNode(
                        meshName,
                        target,
                        new AwaitEvidenceReq($"readiness-{rid}"))
                    .Timeout(TimeSpan.FromMilliseconds(500))
                    .Async<AwaitEvidenceRes>(cancellationToken);
                return Results.Ok(new { ready = true });
            }
            catch (Exception error) when (
                error is ZLinkFrameworkException
                    or TimeoutException
                    or OperationCanceledException)
            {
                return Results.Ok(new { ready = false });
            }
        });
        app.MapGet("/evidence", (EvidenceStore evidence) => Results.Ok(evidence.Snapshot()));
        return app;
    }
}

File: framework/languages/node/e2e/AutomaticTurnDispatch/Server/Session/session-host-factory.ts
import fs from 'node:fs';
import { Module } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import {
  ZLinkMessageFlowLogMode,
  type ZLinkFrameworkRuntime,
  type ZLinkEntrySpot,
  type ZLinkEntrySpotContext,
  type ZLinkMessage,
  type ZLinkSpotActorJoinResult
} from '@zlink-systems/framework';
import { ZLinkRedisLocationStore } from '@zlink-systems/framework-locations-redis';
import { ZLinkModule, ZLINK_FRAMEWORK_RUNTIME, zlinkFramework } from '@zlink-systems/nestjs';
import { AutomaticTurnDispatchNames } from '../../Shared/messages';
import { EvidenceStore } from './Support/evidence-store';
import { closeHttpServer, startHttpServer } from './Support/http-server';
import { createAutomaticTurnConfigurationModule } from '../../configuration';
import { SESSION_OPTIONS, validateSessionOptions } from './Configuration/session-options';
import type { SessionOptions } from './Configuration/session-options';
import { AwaitSessionFactory } from './Handlers/await-session';

class AwaitSessionEntrySpot implements ZLinkEntrySpot {
  readonly context!: ZLinkEntrySpotContext;

  async onActorJoin(actorId: string, request: ZLinkMessage): Promise<ZLinkSpotActorJoinResult> {
    void actorId;
    void request;
    return { accepted: true };
  }

  async onJoinedActor(actor: import('@zlink-systems/framework').ZLinkActor): Promise<void> { void actor; }

  async onLeaveActor(actor: import('@zlink-systems/framework').ZLinkActor): Promise<void> { void actor; }

  async onDisconnectActor(actor: import('@zlink-systems/framework').ZLinkActor): Promise<void> { void actor; }
}

export async function startSessionHost(): Promise<void> {
  let stopping = false;
  const configuration = createAutomaticTurnConfigurationModule(SESSION_OPTIONS, validateSessionOptions);

  class SessionModule {}
  Module({
    imports: [
      configuration,
      ZLinkModule.forRootFactory({
        imports: [configuration],
        inject: [SESSION_OPTIONS],
        useFactory: (value: unknown) => {
          const options = value as SessionOptions;
          fs.mkdirSync(options.logDir, { recursive: true });
          const locationStore = new ZLinkRedisLocationStore({
            url: `redis://${options.redisEndpoint}`,
            keyPrefix: options.redisKeyPrefix
          });
          const builder = zlinkFramework();
          builder
            .addLocationStore(locationStore)
            .configureDispatch()
              .messageFlow(ZLinkMessageFlowLogMode.KeyTransitions)
              .traceLogFile(`${options.logDir}/${options.rid}-flow.log`)
              .traceLabel(options.rid);

          const controlMesh = builder.addRouteMesh(AutomaticTurnDispatchNames.controlChannel)
            .listen(options.controlRouterEndpoint)
            .routingId(options.rid);
          controlMesh.channel(AutomaticTurnDispatchNames.controlChannel).server();
          for (const endpoint of options.playControlEndpoints) controlMesh.peerConnections().connect(endpoint);
          const routeMesh = builder.addRouteMesh(AutomaticTurnDispatchNames.spotRouteChannel)
            .listen(options.spotRouteEndpoint)
            .routingId(options.rid);
          routeMesh.channel(AutomaticTurnDispatchNames.spotRouteChannel).server();
          for (const endpoint of options.playSpotRouteEndpoints) routeMesh.peerConnections().connect(endpoint);
          const spotMesh = builder.addRouteMesh(AutomaticTurnDispatchNames.spotChannel)
            .routingId(options.rid)
            .listen(options.spotRouterEndpoint);
          spotMesh.objects().server().addEntrySpot(AwaitSessionEntrySpot);
          spotMesh.channel(AutomaticTurnDispatchNames.spotChannel).server();
          for (const peer of options.spotRouterPeers) spotMesh.peerConnections().connect(peer.rid, peer.endpoint);
          builder.addStreamNode(AutomaticTurnDispatchNames.streamNode)
            .bind(options.streamEndpoint)
            .enableActorDispatch()
            .registerSession(AwaitSessionFactory);

          return builder.build();
        }
      })
    ],
    providers: [
      {
        provide: EvidenceStore,
        inject: [SESSION_OPTIONS],
        useFactory: (options: SessionOptions) => new EvidenceStore(options.rid, options.evidenceFile)
      },
      AwaitSessionEntrySpot,
      AwaitSessionFactory
    ]
  })(SessionModule);

  const app = await NestFactory.createApplicationContext(SessionModule, { logger: false, abortOnError: false });
  const options = app.get<SessionOptions>(SESSION_OPTIONS);
  const evidence = app.get(EvidenceStore);
  const frameworkRuntime = app.get<ZLinkFrameworkRuntime>(ZLINK_FRAMEWORK_RUNTIME, { strict: false });
  let shutdownOperation: Promise<unknown> | undefined;
  let shutdownError: unknown;
  const server = await startHttpServer(options.httpUrl, [
    {
      method: 'GET',
      path: '/health',
      handle: () => ({
        status: frameworkRuntime.status.acceptingWork ? 'ready' : 'draining',
        role: 'session',
        rid: options.rid
      })
    },
    { method: 'GET', path: '/evidence', handle: () => evidence.snapshot() },
    {
      method: 'GET',
      path: '/status',
      handle: () => ({
        state: frameworkRuntime.status.state,
        acceptingWork: frameworkRuntime.status.acceptingWork,
        sequence: frameworkRuntime.status.sequence.toString()
      })
    },
    {
      method: 'POST',
      path: '/shutdown',
      handle: () => {
        if (shutdownOperation === undefined) {
          shutdownOperation = frameworkRuntime.shutdown({ deadlineMs: 30_000 });
          void shutdownOperation.then(
            () => { stopping = true; },
            (error: unknown) => { shutdownError = error; stopping = true; }
          );
        }
        return {
          status: 'draining',
          acceptingWork: frameworkRuntime.status.acceptingWork,
          state: frameworkRuntime.status.state
        };
      }
    }
  ]);
  while (!stopping) {
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  await closeHttpServer(server);
  if (shutdownError !== undefined) throw shutdownError;
  await app.close();
}
